### PR TITLE
CDAP-4591: New Metadata storage and search

### PR DESCRIPTION
**Pull request by rohitsinha54**
- Implementation of new Metadata Storage and Search https://issues.cask.co/browse/CDAP-4591
- Added more unit test for metadata search to cover different cases
- Fixed bug in tag search: https://issues.cask.co/browse/CDAP-4393
- Fixed bug in mixed case metadata storage and search https://issues.cask.co/browse/CDAP-4331

Design Doc: https://wiki.cask.co/display/CE/Improvements+to+Metadata+Search+and+System+Metadata

Build: http://builds.cask.co/browse/CDAP-RBT584-4

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java [hunk 1]
```diff
@@ -45,6 +45,12 @@
     .or(CharMatcher.is('_')
           .or(CharMatcher.is('-')));
 
+  private static final CharMatcher valueMatcher = CharMatcher.inRange('A', 'Z')
+    .or(CharMatcher.inRange('a', 'z'))
+    .or(CharMatcher.inRange('0', '9'))
+    .or(CharMatcher.is('_')
+          .or(CharMatcher.is('-')).or(CharMatcher.WHITESPACE));
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: nit: indentation. Also fix line 46 please

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java [hunk 2]
```diff
@@ -45,6 +45,13 @@
     .or(CharMatcher.is('_')
           .or(CharMatcher.is('-')));
 
+  private static final CharMatcher valueMatcher = CharMatcher.inRange('A', 'Z')
+    .or(CharMatcher.inRange('a', 'z'))
+    .or(CharMatcher.inRange('0', '9'))
+    .or(CharMatcher.is('_')
+          .or(CharMatcher.is('-'))
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: nit: indentation on lines 46, 52 and 53

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java [hunk 3]
```diff
@@ -198,6 +205,15 @@ private void validateAllowedFormat(Id.NamespacedId entityId, String keyword) thr
   }
 
   /**
+   * Validate the value of a property matches the {@link #valueMatcher} character test.
+   */
+  private void validateAllowedValueFormat(Id.NamespacedId entityId, String keyword) throws InvalidMetadataException {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: nit: better to call these ``validateKeyFormat`` and ``validateValueFormat``?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 4]
```diff
@@ -119,17 +121,39 @@ public void testProperties() throws Exception {
...
     Set<MetadataSearchResultRecord> expected = ImmutableSet.of(
+      new MetadataSearchResultRecord(application)
+    );
+    Set<MetadataSearchResultRecord> searchProperties = searchMetadata(Id.Namespace.DEFAULT,
+                                                                      "aKey:aValue",
+                                                                      MetadataSearchTargetType.APP);
+    Assert.assertEquals(expected, searchProperties);
+    searchProperties = searchMetadata(Id.Namespace.DEFAULT, "multiword:wow1", MetadataSearchTargetType.APP);
+    Assert.assertEquals(expected, searchProperties);
+    searchProperties = searchMetadata(Id.Namespace.DEFAULT, "multiword:woW5", MetadataSearchTargetType.APP);
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: also add a verification for the query 'wow5' since that would verify both case-insensitivity and word search?

Comment (rohitsinha54) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: Its later I changed an existing one

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java [hunk 5]
```diff
@@ -22,14 +22,17 @@
 import javax.annotation.Nullable;
 
 /**
- * Key used to store metadata values.
+ * Key used to store metadata values and indexes.
  */
-public class MdsValueKey {
-  private static final byte[] ROW_PREFIX = {'v'};
+public class MdsKey {
+  private static final byte[] VALUE_ROW_PREFIX = {'v'}; // value row to store metadata value
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java]: nit: the comment should say 'prefix' too right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java [hunk 6]
```diff
@@ -70,28 +73,52 @@ public static String getMetadataKey(String type, byte[] rowKey) {
...
 
-  public static MDSKey getMDSKey(Id.NamespacedId targetId, @Nullable String key) {
-    String targetType = KeyHelper.getTargetType(targetId);
-    MDSKey.Builder builder = new MDSKey.Builder();
-    builder.add(ROW_PREFIX);
-    builder.add(targetType);
-    KeyHelper.addNamespaceIdToKey(builder, targetId);
+  /**
+   * Creates a key for metadata value row in the format:
+   * [{@link #VALUE_ROW_PREFIX}][targetType][targetId][key] for value index rowss
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java]: typo: rows

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java [hunk 7]
```diff
@@ -70,28 +73,52 @@ public static String getMetadataKey(String type, byte[] rowKey) {
...
+  public static MDSKey getMDSIndexKey(Id.NamespacedId targetId, String key, @Nullable String index) {
+    MDSKey.Builder builder = getMDSKeyPrefix(targetId, INDEX_ROW_PREFIX);
+    builder.add(key);
+    if (index != null) {
+      builder.add(index);
+    }
     return builder.build();
   }
 
+  private static MDSKey.Builder getMDSKeyPrefix(Id.NamespacedId targetId, byte[] indexRowPrefix) {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java]: the second param should just be ``rowPrefix``, as it can also be a prefix for the value row?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java [hunk 8]
```diff
@@ -70,28 +73,52 @@ public static String getMetadataKey(String type, byte[] rowKey) {
...
+    String targetType = KeyHelper.getTargetType(targetId);
+    MDSKey.Builder builder = new MDSKey.Builder();
+    builder.add(indexRowPrefix);
+    builder.add(targetType);
+    KeyHelper.addNamespaceIdToKey(builder, targetId);
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java]: not your PR, but this method name is confusing, since it seems to indicate that it adds a namespace id to the key, when it actually adds an ``Id.NamespacedId``. Better to call it ``addEntityIdToKey``?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java [hunk 9]
```diff
@@ -70,28 +73,52 @@ public static String getMetadataKey(String type, byte[] rowKey) {
...
+    return builder;
+  }
+
   public static Id.NamespacedId getNamespaceIdFromKey(String type, byte[] rowKey) {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java]: ditto

File: cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchTargetType.java [hunk 10]
```diff
@@ -29,11 +31,28 @@
...
+
+  @Override
+  public String toString() {
+    return getValue();
+  }
+
+  /**
+   * @return {@link MetadataSearchTargetType} of the given value.
+   */
+  public static MetadataSearchTargetType getEnum(String value) {
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchTargetType.java]: enum methods are generally called ``valueOf*``. Better to call this ``valueOfInternalName`` or ``valueOfSerializedForm``

File: cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchTargetType.java [hunk 11]
```diff
@@ -29,11 +31,28 @@
 
   private final String internalName;
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchTargetType.java]: perhaps call this ``serializedForm`` to better indicate what it is?

File: cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchTargetType.java [hunk 12]
```diff
@@ -29,11 +31,28 @@
...
 
-  private MetadataSearchTargetType(String internalName) {
+  MetadataSearchTargetType(String internalName) {
     this.internalName = internalName;
   }
 
-  public String getInternalName() {
+  public String getValue() {
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchTargetType.java]: this is a getter, so lets have it in the form ``get<MemberName>``

File: cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchTargetType.java [hunk 13]
```diff
@@ -29,11 +31,28 @@
...
+  /**
+   * @return {@link MetadataSearchTargetType} of the given value.
+   */
+  public static MetadataSearchTargetType getEnum(String value) {
+    for (MetadataSearchTargetType metadataSearchTargetType : values()) {
+      if (metadataSearchTargetType.getValue().equalsIgnoreCase(value)) {
+        return metadataSearchTargetType;
+      }
+    }
+    throw new IllegalArgumentException("Enum not found for value: " + value);
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchTargetType.java]: nit: should say ``String.format("No enum constant co.cask.cdap.proto.metadata.MetadataSearchTargetType.%s", value)`` like standard enum

Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchTargetType.java]: actually I think this should be ``"No enum constant for serialized form: " + value``

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 14]
```diff
@@ -53,12 +54,12 @@
     .registerTypeAdapter(Id.NamespacedId.class, new NamespacedIdCodec())
     .create();
 
-  private static final String HISTORY_COLUMN = "h";
-  private static final String VALUE_COLUMN = "v";
-  private static final String TAGS_SEPARATOR = ",";
+  private static final Pattern VALUE_SPLIT_PATTERN = Pattern.compile("\\s*[-_,\\s]\\s*");
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: should we add a comment explaining the reasoning behind this pattern and the logic of splitting?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 15]
```diff
@@ -53,12 +54,12 @@
...
 
-  static final String KEYVALUE_COLUMN = "kv";
-  static final String CASE_INSENSITIVE_VALUE_COLUMN = "civ";
+  private static final String HISTORY_COLUMN = "h"; // column for metadata history
+  private static final String VALUE_COLUMN = "vc";  // column for metadata value
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: just ``h``, ``v``, and ``i`` as column names for consistency?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 16]
```diff
@@ -53,12 +54,12 @@
...
+  static final String INDEX_COLUMN = "ic";          // column for metadata indexes
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: nit: better to have this after the next constant to keep all the private members together?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 17]
```diff
@@ -270,6 +273,31 @@ private void removeMetadata(Id.NamespacedId targetId, Predicate<String> filter)
   }
 
   /**
+   * Deletes all indexes associated with a metadata key
+   * @param targetId the {@link Id.NamespacedId} for which keys are to be removed
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: nit: space above this line

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 18]
```diff
@@ -270,6 +273,31 @@ private void removeMetadata(Id.NamespacedId targetId, Predicate<String> filter)
...
+   * @param metadataKey the key to remove from the metadata of the specified {@link Id.NamespacedId}
+   */
+  private void deleteIndexes(Id.NamespacedId targetId, String metadataKey) {
+    MDSKey mdsKey = MdsKey.getMDSIndexKey(targetId, metadataKey, null);
+    byte[] prefix = mdsKey.getKey();
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: this is a ``startKey``, right, not ``prefix``?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 19]
```diff
@@ -370,117 +376,98 @@ private Metadata getSnapshotBeforeTime(Id.NamespacedId targetId, long timeMillis
     }
   }
 
-  // Helper method to execute IndexedTable search on target index column.
-  private List<MetadataEntry> executeSearch(final String namespaceId, final String column,
-                                            final String searchValue, final MetadataSearchTargetType type) {
-    List<MetadataEntry> results = new LinkedList<>();
-
+  public List<MetadataEntry> search(final String namespaceId, final String searchValue,
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: should add a javadoc for public methods

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 20]
```diff
@@ -370,117 +376,98 @@ private Metadata getSnapshotBeforeTime(Id.NamespacedId targetId, long timeMillis
...
   /**
+   * Store indexes for a {@link MetadataEntry}
+   *
+   * @param targetId the {@link Id.NamespacedId} from which the metadata indexes has to be stored
+   * @param entry the {@link MetadataEntry} which has to be indexed
+   */
+  private void storeIndexes(Id.NamespacedId targetId, MetadataEntry entry) {
+    // create a list of indexes by splitting on whitespaces and comma
+    Set<String> indexes = Sets.newHashSet(Arrays.asList(VALUE_SPLIT_PATTERN.split(entry.getValue())));
+    indexes.add(entry.getValue()); // add the complete metadata value as an index too
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: this should be added only if it does not exist, right?

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java [hunk 21]
```diff
@@ -201,26 +201,38 @@ public void testSearchOnTags() throws Exception {
     Assert.assertEquals(0, dataset.getTags(stream1).size());
     dataset.addTags(app1, "tag1", "tag2", "tag3");
     dataset.addTags(flow1, "tag1");
-    dataset.addTags(dataset1, "tag3", "tag2", "tag12");
+    dataset.addTags(dataset1, "tag3", "tag2", "tag12-tag33");
```
Comment (bdmogal) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java]: didn't we decide that we shouldn't allow splitting for tags? I mean tag12 and tag33 could easily be two separate tags?

Comment (bdmogal) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java]: just remembered that we probably spoke about not allowing spaces. but I think similarly, we shouldn't do any splitting at all for tags. If the tag is tag12-tag33, then *tag33, tag12*, tag12-tag33 should match but maybe not tag33? What do you think?

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java [hunk 22]
```diff
@@ -201,26 +201,38 @@ public void testSearchOnTags() throws Exception {
...
+    // Try to search for tag33
+    results = dataset.search("ns1", "tags:tag33", MetadataSearchTargetType.ALL);
+    Assert.assertEquals(1, results.size());
+
+    // Try to search for tag33
+    results = dataset.search("ns1", "tag33", MetadataSearchTargetType.ALL);
+    Assert.assertEquals(1, results.size());
+
+    // Try to search for tag3
+    results = dataset.search("ns1", "tags:tag3*", MetadataSearchTargetType.ALL);
```
Comment (bdmogal) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java]: perhaps add some verifications for a different namespace too, just for completeness

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java [hunk 23]
```diff
@@ -234,39 +246,66 @@ public void testSearchOnTags() throws Exception {
...
-      dataset.searchByValue("ns1", "value1", MetadataSearchTargetType.PROGRAM);
+      dataset.search("ns1", "value1", MetadataSearchTargetType.PROGRAM);
 
     // Assert check
     Assert.assertEquals(1, results.size());
 
     MetadataEntry result = results.get(0);
-    Assert.assertEquals(record, result);
+    Assert.assertEquals(entry, result);
+
```
Comment (bdmogal) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java]: nit: extra newline

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java [hunk 24]
```diff
@@ -234,39 +246,66 @@ public void testSearchOnTags() throws Exception {
 
   @Test
   public void testSearchOnValue() throws Exception {
-    // Create record
-    MetadataEntry record = new MetadataEntry(flow1, "key1", "value1");
+    // Create entry
+    MetadataEntry entry = new MetadataEntry(flow1, "key1", "value1");
+    String multiWordValue = "aV1 av2 ,  -  ,  av3 - av4_av5 av6";
```
Comment (bdmogal) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java]: what happens for the search queries ``-`` or ``_`` or ``,`` ?

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java [hunk 25]
```diff
@@ -279,7 +318,7 @@ public void testSearchOnValue() throws Exception {
     dataset.setProperty(stream1, "key21", "value21");
 
     // Search for it based on value asterix
-    List<MetadataEntry> results3 = dataset.searchByValue("ns1", "value2*",
+    List<MetadataEntry> results3 = dataset.search("ns1", "value2*",
                                                          MetadataSearchTargetType.ALL);
```
Comment (bdmogal) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java]: nit: indentation

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java [hunk 26]
```diff
@@ -300,26 +339,51 @@ public void testSearchOnValue() throws Exception {
...
+    // Search for it based on a word in value
+    results =
+      dataset.search("ns1", "multiword" + MetadataDataset.KEYVALUE_SEPARATOR + "aV5",
+                     MetadataSearchTargetType.PROGRAM);
+
+    // Assert check
+    Assert.assertEquals(1, results.size());
+    result = results.get(0);
+    Assert.assertEquals(multiWordEntry, result);
+
```
Comment (bdmogal) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java]: nit: extra newline

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java (hunk 21 above)
Comment (rohitsinha54) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java]: I agree we shouldn't do any splitting for tags besides on tag separator to get individual tags.
@poornachandra : Thoughts ?

**Comment by bdmogal** [comment]
@rohitsinha54: Other than a couple of comments, the rest are pretty minor from me. Apart from them, this looks good. However, please get this reviewed with @poornachandra too.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java [hunk 27]
```diff
@@ -70,28 +73,52 @@ public static String getMetadataKey(String type, byte[] rowKey) {
...
     MDSKey.Splitter keySplitter = new MDSKey(rowKey).split();
 
-    // The rowkey is [rowPrefix][targetType][targetId][metadata-type][key], so skip the first two.
+    // The rowkey is [rowPrefix][targetType][targetId][key] for value rows and
+    // [rowPrefix][targetType][targetId][key][index] for value index rows
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java]: It will be easier to read if you align the key values

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java [hunk 28]
```diff
@@ -100,7 +127,9 @@ public static MDSKey getMDSKey(Id.NamespacedId targetId, @Nullable String key) {
   public static String getNamespaceId(MDSKey key) {
     MDSKey.Splitter keySplitter = key.split();
 
-    // The rowkey is [rowPrefix][targetType][targetId][metadata-type][key], so skip the first two.
+    // The rowkey is [rowPrefix][targetType][targetId][key] for value rows and
+    // [rowPrefix][targetType][targetId][key][index] for value index rows
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java]: Same here - alignment for readability

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java (hunk 20 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: ignore, missed that its a set

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java (hunk 21 above)
Comment (poornachandra) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java]: Only issue then is we'll have different ways of matching for tags and for other metadata entities. This might lead to confusion.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java (hunk 14 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: Can't this pattern just be `[-_,\\s]*`?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java [hunk 29]
```diff
@@ -45,6 +45,13 @@
     .or(CharMatcher.is('_')
           .or(CharMatcher.is('-')));
 
+  private static final CharMatcher valueMatcher = CharMatcher.inRange('A', 'Z')
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: static final field name should be ALL_CAPS, right? Looks like checkstyle flags only if field is public.

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 30]
```diff
@@ -150,6 +174,16 @@ public void testProperties() throws Exception {
     );
     Assert.assertEquals(expected, searchProperties);
 
+    searchProperties = searchMetadata(Id.Namespace.DEFAULT, "multiword:w*", MetadataSearchTargetType.ALL);
+    Assert.assertEquals(2, searchProperties.size());
```
Comment (poornachandra) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: It would be a good idea to assert the exact ids too, not just size. Same below too.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 31]
```diff
@@ -253,13 +254,15 @@ private void removeMetadata(Id.NamespacedId targetId, Predicate<String> filter)
...
-        if (keyValue == null && value == null) {
+        if (value == null) {
           continue;
         }
-        if (filter.apply(MdsValueKey.getMetadataKey(targetType, next.getRow()))) {
+        String metadataKey = MdsKey.getMetadataKey(targetType, next.getRow());
+        if (filter.apply(metadataKey)) {
           indexedTable.delete(new Delete(next.getRow()));
+          // delete search indexes
+          deleteIndexes(targetId, metadataKey);
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: This will lead to nested scans. Can we not do it using a single scan?

**Comment by poornachandra** [comment]
@rohitsinha54 Overall the code looks good. It has also become much simpler. It needs some more test cases. Let's discuss tomorrow.

**Comment by rohitsinha54** [comment]
Yes. I agree it needs more unit tests due to all possible edge cases. Specially unit test which confirm what all indexes were created. But I was asked to open the PR and get initial feedback. We can discuss today and add some more.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java (hunk 7 above)
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java]: Intellij refactor :/

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java (hunk 24 above)
Comment (rohitsinha54) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java]: Nothing gets matched. Added unit tests for it

**Comment by poornachandra** [comment]
LGTM

**Comment by rohitsinha54** [comment]
Build passed here: http://builds.cask.co/browse/CDAP-RBT584-10
Merging.
